docs/spec/fastsync/fastsync.md

```diff
@@ -0,0 +1,1164 @@
+specify below under what (favorable) conditions *Fastsync* ensures to
+terminate successfully, and satisfy the requirements of the sequential
+problem statement:
+
+#### **[FS-DISTR-TERM]**:
+*Fastsync* eventually terminates: it either *terminates successfully* or it  *terminates with failure*.
+
+
+#### Remote Functions
+
+The Tendermint Full Node exposes the following functions over
+remote procedure calls. The "Expected precondition" are only expected for
+correct peers (as no assumption is made on internals of faulty
+processes [FS-A-PEER]).
+
+
+> In this document we describe the communication with peers 
+via asynchronous RPCs.
+
+
+```go
+func Status(addr Address) (int64, error)
+```
+- Implementation remark
+   - RPC to full node *addr*
+- Expected precondition
+  - none
+- Expected postcondition
+  - if *addr* is correct: Returns the current height `height` of the
+    peer. [FS-A-COMM]
+  - if *addr* is faulty: Returns an arbitrary height. [**[TMBC-AUTH-BYZ]**][TMBC-Auth-Byz-link]
+- Error condition
+   * if *addr* is correct: none. By [FS-A-COMM] we assume communication is reliable and timely.
+   * if *addr* is faulty: arbitrary error (including timeout). [**[TMBC-AUTH-BYZ]**][TMBC-Auth-Byz-link]
+----
+
+
+ ```go
+func Block(addr Address, height int64) (Block, error)
+```
+- Implementation remark
+   - RPC to full node *addr*
+- Expected precondition
+  - 'height` is less than or equal to height of the peer
+- Expected postcondition
+  - if *addr* is correct: Returns the block of height `height`
+  from the blockchain. [FS-A-COMM]
+  - if *addr* is faulty: Returns arbitrary block [**[TMBC-AUTH-BYZ]**][TMBC-Auth-Byz-link]
+- Error condition
+  - if *addr* is correct: precondition violated. [FS-A-COMM]
+  - if *addr* is faulty: arbitrary error (including timeout). [**[TMBC-AUTH-BYZ]**][TMBC-Auth-Byz-link]
+----
+
+### Temporal Properties
+
+
+#### Fairness
+
+As mentioned above, without assumptions on the correctness of some
+peers, no protocol can achieve the required specifications. Therefore, 
+we sometimes consider the following (fairness) constraint in the
+safety and liveness properties below:
+
+
+#### **[FS-SOME-CORR-PEER]**:
+Initially, the set *peerIDs* contains at least one correct full node.
+
+> While in principle the above condition can be part of a sufficient
+> condition to solve [FS-SEQ-LIVE] and
+> [FS-SEQ-TERM] and [FS-SEQ-TERM-SYNC], we will see below that the
+> current implementation of Fastsync (V2) requires the following (much
+> stronger) requirement
+
+#### **[FS-ALL-CORR-PEER]**:
+At all times, the set *peerIDs* contains only correct full nodes.
+
+
+#### Safety
+
+
+<!--
+> safety specifications / invariants in English 
+---->
+
+
+#### **[FS-VC-ALL-CORR-NONABORT]**:
+Under [FS-ALL-CORR-PEER], *Fastsync* never terminates with failure.
+
+
+#### **[FS-VC-STATE-INV]**:
+If *FastSync* terminates successfully at height *terminationHeight*, then the
```

Comment:
Isn't this always the case? Difference between success and failure in my view is only with respect to the quality of the result, i.e, about number of synced blocks.

Reply:
I wanted to recall *terminationheight* here, but we can do without, I removed it.